pyfar/signals/deterministic.py

```diff
@@ -256,6 +389,478 @@ def exponential_sweep_time(n_samples, frequency_range, n_fade_out=90,
+        >>> sweep, _ = pf.signals.magnitude_spectrum_weighted_sweep(
+        ...     2**16, magnitude, 5000, 1000)
+        >>> pf.plot.time_freq(sweep)
+    """
+
+    signal, group_delay = _frequency_domain_sweep(
+        n_samples=n_samples,
+        sweep_type=magnitude_spectrum,
+        frequency_range=[0, sampling_rate / 2],
+        butterworth_order=0,
+        start_margin=start_margin,
+        stop_margin=stop_margin,
+        fade_in=fade_in,
+        fade_out=fade_out,
+        sampling_rate=sampling_rate)
+
+    return signal, group_delay
+
+
+def linear_perfect_sweep(n_samples, sampling_rate=44100):
+    """
+    Generate a perfect linear sweep in the frequency domain.
+
+    The perfect sweep is generated according to [#]_ and is used for adaptive
+    system identification. It is orthogonal to delayed versions of itself and
+    can be played back in a loop.
+
+    Parameters
+    ----------
+    n_samples : int
+        The length of the sweep in samples.
+    sampling_rate : int
+        The sampling rate in Hz. The default is ``44100``.
+
+    Returns
+    -------
+    sweep : Signal
+        The sweep signal. The Signal is in the time domain, has a maximum
+        absolute amplitude of 1 and the ``rms`` FFT normalization
+        (see :py:func:`~pyfar.dsp.fft.normalization`).
+    group_delay_sweep : FrequencyData
+        The group delay of the sweep in seconds as a single sided spectrum.
+
+    References
+    ----------
+    .. [#] C. Antweiler, A. Telle, P. Vary, and G. Enzner, “Perfect-sweep NLMS
+           for time-variant acoustic system identification,” in IEEE
+           Int. Conf. Acoustics, Speech and Signal Processing (ICASSP), Prague,
+           Czech Republic, May 2011. doi: 10.1109/ICASSP.2012.6287930.
+
+    Examples
+    --------
+    Plot a shifted perfect sweep to show that it can be looped
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>> sweep, _ = pf.signals.linear_perfect_sweep(2**8)
+        >>> sweep_shifted = pf.dsp.time_shift(sweep, 2**7)
+        >>> ax = pf.plot.time_freq(sweep_shifted, unit='samples')
+        >>> ax[0].axvline(2**7, ls='--', c='r', label='start/end')
+        >>> ax[0].legend(loc=4)
+
+    Show that the sweep is orthogonal to all shifted version of itself. This
+    property is important for adaptive system identification.
+
+    .. plot::
+
+        >>> import pyfar as pf
+        >>> import numpy as np
+        >>> import matplotlib.pyplot as plt
+        >>>
+        >>> sweep, _ = pf.signals.linear_perfect_sweep(2**8)
+        >>>
+        >>> # compute auto-correlation
+        >>> auto_correlation = np.empty(2**8)
+        >>> for idx, shift in enumerate(range(-2**7, 2**7)):
+        >>>     auto_correlation[idx] = np.dot(
+        ...         sweep.time.flatten(),
+        ...         np.roll(sweep.time.flatten(), shift))
+        >>>
+        >>> auto_correlation /= pf.dsp.energy(sweep)
```

Comment:
If I remember correctly, alternative methods to calculate the autocorrelation are the convolution with the time-flipped (real-valued) signal, or via the inverse fourier transform of the squared magnitude spectrum. Maybe these approaches lead to shorter code using pyfar.dsp functionality?

Reply:
Indeed, this can be done in a more compact way and especially without a loop.  For example using https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.correlate.html

Reply:
I tried this, but its not cyclic (regardless of `mode` and `method`) and causes ringing in the auto-correlation:

```python
import pyfar as pf
import numpy as np
import scipy.signal as sgn
import matplotlib.pyplot as plt
%matplotlib inline

sweep = pf.signals.linear_perfect_sweep(2**8)
# compute auto-correlation
auto_correlation = sgn.correlate(
    sweep.time.flatten(), sweep.time.flatten())
auto_correlation /= pf.dsp.energy(sweep)
# plot auto-correlation
with pf.plot.context():
    plt.plot(np.arange(-2**7, 2**7), auto_correlation)
    plt.gca().set_xlim(-2**7, 2**7)
    plt.gca().set_xlabel('time lag in samples')
    plt.gca().set_ylabel('auto correlation')
```

I'm this leaving this as it was...

Reply:
Mh, this is a peculiar issue with the integral approximation in SciPy I'd assume...